### PR TITLE
Region permissions

### PR DESCRIPTION
**Pull request by intOrfloat**
Haven't thoroughly tested or removed excess console.log's

### CODE REVIEW

File: client/scripts/DrawTogether.js [hunk 1]
```diff
@@ -597,8 +599,16 @@ DrawTogether.prototype.changeRoom = function changeRoom (room, number, x, y, spe
 			this.paint.addPublicDrawings(this.decodeDrawings(drawings));
 			this.chat.addMessage("Invite people", "http://www.anondraw.com/#" + room + number);
 			
-			if (this.account.uKey) {
```
Comment (Squarific) [inline client/scripts/DrawTogether.js]: Why did you change this?

File: client/scripts/DrawTogether.js [hunk 2]
```diff
@@ -11,6 +11,7 @@ function DrawTogether (container, settings) {
 	this.playerList = [];
 	this.moveQueue = [];
 	this.favList = [];
+	this.regionsList = []; // Specific to the user
```
Comment (Squarific) [inline client/scripts/DrawTogether.js]: Maybe name this "myRegions" ?

File: client/scripts/DrawTogether.js [hunk 3]
```diff
@@ -638,7 +649,17 @@ DrawTogether.prototype.joinGame = function joinGame () {
 			this.chat.addMessage("Welcome to anondraw, enjoy your game!");
 			this.chat.addMessage("Invite friends:", "http://www.anondraw.com/#" + room);
 			
-			this.getFavorites();
```
Comment (Squarific) [inline client/scripts/DrawTogether.js]: This is not a good way to do it. Randomly waiting gives very very very very very big problems down the line.
If you want to do something when the server knows our rep; just put it in the event handler where we get our reputation.

File: client/scripts/DrawTogether.js [hunk 4]
```diff
@@ -596,8 +599,16 @@ DrawTogether.prototype.changeRoom = function changeRoom (room, number, x, y, spe
 			this.paint.addPublicDrawings(this.decodeDrawings(drawings));
 			this.chat.addMessage("Invite people", "http://www.anondraw.com/#" + room + number);
 			
-			if (this.account.uKey) {
+			if (typeof this.reputation === 'undefined') {//not logged in
+				setTimeout(function(){
```
Comment (Squarific) [inline client/scripts/DrawTogether.js]: Copy pasted from the other one:

This is not a good way to do it. Randomly waiting gives very very very very very big problems down the line.
If you want to do something when the server knows our rep; just put it in the event handler where we get our reputation.

File: client/scripts/DrawTogether.js [hunk 5]
```diff
@@ -1609,18 +1718,37 @@ DrawTogether.prototype.createFavorite = function (x, y, name) {
...
+	var regionCount = this.regionsList.length || 0;
+
+	if (!this.memberlevel) {
+		if(this.reputation < this.REGION_MIN_REP){
+			
+			this.chat.addMessage("You must have at least "+ this.REGION_MIN_REP +" rep!");
+			return;
+		}
+		if(typeof this.regionsList !== 'undefined')
+		{
```
Comment (Squarific) [inline client/scripts/DrawTogether.js]: Curcly open brackets should be on the same line

File: client/scripts/DrawTogether.js [hunk 6]
```diff
@@ -1609,18 +1718,37 @@ DrawTogether.prototype.createFavorite = function (x, y, name) {
...
-	if (!this.memberlevel) {
-		this.chat.addMessage("Creating a protected region is a member only feature.");
+	if (typeof this.reputation === 'undefined') { 
+		this.chat.addMessage("You must be logged in to create protected regions.");
 		return;
 	}
+	var regionCount = this.regionsList.length || 0;
+
+	if (!this.memberlevel) {
+		if(this.reputation < this.REGION_MIN_REP){
```
Comment (Squarific) [inline client/scripts/DrawTogether.js]: There should be a space after if and one before the curly bracket.

File: client/scripts/DrawTogether.js [hunk 7]
```diff
@@ -1609,18 +1718,37 @@ DrawTogether.prototype.createFavorite = function (x, y, name) {
...
+			
+			this.chat.addMessage("You must have at least "+ this.REGION_MIN_REP +" rep!");
+			return;
+		}
+		if(typeof this.regionsList !== 'undefined')
+		{
+			if(typeof this.regionsList.length === 'number')
```
Comment (Squarific) [inline client/scripts/DrawTogether.js]: If whatever you are doing the if for is more than one line then I don't like this style. You should include the brackets.

File: client/scripts/DrawTogether.js [hunk 8]
```diff
@@ -1632,11 +1760,87 @@ DrawTogether.prototype.resetProtectedRegions = function () {
...
+		
+		this.chat.addMessage("Regions", result.success);
+		console.log("result:",result);
+		//this.regionsList = JSON.parse(result);
+	}.bind(this));
+};
+
+DrawTogether.prototype.removeUsersToMyProtectedRegion = function (userIdArr, regionId, callback) {
+	console.log("userIdArr", userIdArr);
+	this.network.socket.emit("removeuserstomyprotectedregion", userIdArr, regionId, function (err, result) {
```
Comment (Squarific) [inline client/scripts/DrawTogether.js]: In english it would be remove users FROM my protected region so even if this might sound more consistent I think keeping it readable might be better.

These are also very long so they are a bit hard to read. Maybe somehting like "protectedregion.removeuser" ? 
Or maybe do it camelCase?

File: server/playerServer/database.sql [hunk 9]
```diff
@@ -47,8 +47,16 @@ CREATE TABLE IF NOT EXISTS ipbans (
     INDEX (ip, enddate)
 );
 
+--alter table regions add id INT UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY FIRST; 
+--ALTER TABLE regions ADD minRepAllowed int UNSIGNED
+
+--Need above for the production server when permissions are done^^
+
 CREATE TABLE IF NOT EXISTS regions (
+    id INT UNSIGNED NOT NULL AUTO_INCREMENT,
```
Comment (Squarific) [inline server/playerServer/database.sql]: Auto increment is default for primary key.

File: server/playerServer/database.sql [hunk 10]
```diff
@@ -58,6 +66,12 @@ CREATE TABLE IF NOT EXISTS regions (
     INDEX (owner, maxX, minX, maxY, minY, room)
 );
 
+CREATE TABLE IF NOT EXISTS regions_permissions (
+    regionId INT UNSIGNED,
+    allowedUser INT UNSIGNED,
```
Comment (Squarific) [inline server/playerServer/database.sql]: I'd rather have this be called userId since in the future we want to add multiple permissiontypes.

File: server/playerServer/scripts/PlayerDatabase.js [hunk 11]
```diff
@@ -350,7 +350,7 @@ PlayerDatabase.prototype.addProtectedRegion = function addProtectedRegion (useri
 				return;
 			}
 
-			this.database.query("INSERT INTO regions (owner, minX, minY, maxX, maxY, room) VALUES (?, ?, ?, ?, ?, ?)", [userid, minX, minY, maxX, maxY, room], function (err) {
+			this.database.query("INSERT INTO regions (owner, minX, minY, maxX, maxY, room, minRepAllowed) VALUES (?, ?, ?, ?, ?, ?, 300)", [userid, minX, minY, maxX, maxY, room], function (err) {
```
Comment (Squarific) [inline server/playerServer/scripts/PlayerDatabase.js]: Random magic number. Put it in a variable at the top of the file like var DEFAULT_MIN_REGION_REP = 300; and make it waaay higher. Like maybe even maxInt or close to it.

File: server/playerServer/scripts/PlayerDatabase.js [hunk 12]
```diff
@@ -363,17 +363,154 @@ PlayerDatabase.prototype.resetProtectedRegions = function resetProtectedRegions
...
+		this.database.query("delete from regions_permissions where regionId = ?", [regionId], function (err) {
+			callback(err);
+		}.bind(this));
+	}.bind(this));
+};
+
 PlayerDatabase.prototype.getProtectedRegions = function getProtectedRegions (room, callback) {
 	// TODO: Select permissions
-	this.database.query("SELECT owner, minX, minY, maxX, maxY FROM regions WHERE room = ?", [room], function (err, rows, fields) {
+	this.database.query("SELECT id, owner, minX, minY, maxX, maxY, minRepAllowed FROM regions WHERE room = ? ORDER BY id", [room], function (err, rows1, fields) {
```
Comment (Squarific) [inline server/playerServer/scripts/PlayerDatabase.js]: We might at some point select the protected regions without the permissions. So this should be split into 2 functions.

File: server/realTime/scripts/Network.js [hunk 13]
```diff
@@ -19,6 +19,7 @@ var BIG_BRUSH_MIN_REP = 5;
 var MEMBER_MIN_REP = 15;
 var UPVOTE_MIN_REP = 7;                  // Has to be changed in the playerserver too
 var SHARE_IP_MIN_REP = MEMBER_MIN_REP;
+var REGION_MIN_REP = 25;
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: Given the fact that there is currently no size limit yet. I think I'd rather have a higher number. Unless we first add restrictions.

File: client/scripts/DrawTogether.js [hunk 14]
```diff
@@ -1609,18 +1718,37 @@ DrawTogether.prototype.createFavorite = function (x, y, name) {
...
+				if(this.regionsList.length > 1)
+				{
+					this.chat.addMessage("Having more than one region is premium only!");
+					return;
+				}
+		}
+	}
 
-	this.network.socket.emit("createprotectedregion", from, to, function (err, result) {
+	this.network.socket.emit("createprotectedregion", from, to, regionCount, function (err, result) {
```
Comment (Squarific) [inline client/scripts/DrawTogether.js]: This is not safe. This does not fix the fact that the amount of regions is checked client side. So we are still trusting the client. If you want to do the premium check server side then getting the amount of regions is also something you have to do server side.

Preferably on the playerserver.

File: server/realTime/scripts/Network.js [hunk 15]
```diff
@@ -1072,6 +1140,130 @@ Protocol.prototype.bindIO = function bindIO () {
 			});
 		});
 
+		socket.on("removeprotectedregion", function (regionId, callback) {
+			if (typeof callback !== 'function') return;
+
+			if(isNaN(regionId))
+			{
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: Not enough spaces and curly bracket on a new line when it shouldn't.

File: server/realTime/scripts/Network.js [hunk 16]
```diff
@@ -1072,6 +1140,130 @@ Protocol.prototype.bindIO = function bindIO () {
...
+				room: socket.room,
+				regionId: regionId
+			}, function (err, data) {
+				callback(err, data);
+				protocol.updateProtectedRegions(socket.room);
+			});
+		});
+
+		socket.on("getmyprotectedregions", function (callback) {
+			if(!socket.userid)
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: Why do you do this trough the realTime server? Also why is there a variable asdf and why does it get stringified?

File: client/scripts/DrawTogether.js (hunk 1 above)
Comment (intOrfloat) [inline client/scripts/DrawTogether.js]: i noticed that it this.account.uKey would exist and you would still not be logged in( like on playerserver restarts ) and that the reputation was always a clear indicator

File: server/realTime/scripts/Network.js (hunk 16 above)
Comment (intOrfloat) [inline server/realTime/scripts/Network.js]: i'm not certain why i stringified it. I use the realtime's `protocol.getUsersProtectedRegions(socket.userid, socket.room)` which uses the local variable `this.protectedRegions[room]`. i figured that using the local variable that has the permissions already would be better than a database hit that would need to be joined between users, permissions, and regions.

File: server/realTime/scripts/Network.js [hunk 17]
```diff
@@ -116,7 +126,15 @@ Protocol.prototype.updateProtectedRegions = function updateProtectedRegions (roo
 				new SAT.Vector(data[k].maxX, data[k].maxY),
 				new SAT.Vector(data[k].minX, data[k].maxY),
 			]);
-
+			for(var f = minRegionId; f < permissions.length; f++){
+				if(permissions[f].regionId == data[k].id){
+					if(typeof data[k].permissions === "undefined")
+						data[k].permissions = new Array();
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: You shouldn't use new Array(). Use the [] syntax.

Comment (Squarific) [inline server/realTime/scripts/Network.js]: You also should just write this as data[k].permissions = data[k].permissions || [] instead of using an if

File: server/realTime/scripts/Network.js [hunk 18]
```diff
@@ -202,29 +220,68 @@ Protocol.prototype.getRegionSearchFromSat = function getRegionSearchFromSat (sat
...
+					maxX: protectedRegionsArr[i].maxX,
+					maxY: protectedRegionsArr[i].maxY,
+					minRepAllowed: protectedRegionsArr[i].minRepAllowed
+				});
+	}
+	return p;
+};
+
+Protocol.prototype.isInsideProtectedRegion = function isInsideProtectedRegion (reputation, user, satObjects, room) {
+	if (!this.protectedRegions[room]) return {isNotAllowed: false};
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: You generally should not write variables or functions with negatives. It should be isAllowed: true, isNotAllowed: false is very hard to understand. ;)

File: server/realTime/scripts/Network.js [hunk 19]
```diff
@@ -202,29 +220,68 @@ Protocol.prototype.getRegionSearchFromSat = function getRegionSearchFromSat (sat
 	}
 };
 
-Protocol.prototype.isInsideProtectedRegion = function isInsideProtectedRegion (owner, satObjects, room) {
+Protocol.prototype.getUsersProtectedRegions = function getUsersProtectedRegions (user, room) {
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: This name is rather unclear. Should probably be renamed to something like getProtectedRegionsOwnedBy

File: server/realTime/scripts/Network.js [hunk 20]
```diff
@@ -202,29 +220,68 @@ Protocol.prototype.getRegionSearchFromSat = function getRegionSearchFromSat (sat
...
 
+	var p = [];
+	var protectedRegionsArr = this.protectedRegions[room].all();
+
+	if(protectedRegionsArr.length == 0) return false;
+
+	for (var i = protectedRegionsArr.length - 1; i > 0; i--) {
+		//if (typeof protectedRegionsArr[i].permissions !== "undefined")
+			if( protectedRegionsArr[i].owner === user)
+				p.push({
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: Why do you copy this?

File: server/realTime/scripts/Network.js [hunk 21]
```diff
@@ -202,29 +220,68 @@ Protocol.prototype.getRegionSearchFromSat = function getRegionSearchFromSat (sat
...
 				if (SAT.testPolygonPolygon(relevantRegions[i].satBox, satObjects[k])) {
-					return true
+					return {isNotAllowed: true, minRepAllowed:relevantRegions[i].minRepAllowed, regionid: relevantRegions[i].id, ownerid: relevantRegions[i].owner};
 				}
 			}
 		}
 	}
 
-	return false;
+	return {isNotAllowed: false};
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: It would be cleaner to just return the conflicting region.

File: server/realTime/scripts/Network.js [hunk 22]
```diff
@@ -1072,6 +1146,125 @@ Protocol.prototype.bindIO = function bindIO () {
...
+			if(!asdf){
+				callback("User has no protected regions.");
+				return;
+			}
+
+			callback(null, asdf);
+		});
+
+		socket.on("adduserstomyprotectedregion", function (userIdArr, regionId, callback) {
+			if(!socket.userid)			{
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: Random spaces

File: server/realTime/scripts/Network.js [hunk 23]
```diff
@@ -1072,6 +1146,125 @@ Protocol.prototype.bindIO = function bindIO () {
...
+				callback("No User");
+				return;
+			}
+			if(!socket.room){
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: No spaces

File: server/realTime/scripts/Network.js [hunk 24]
```diff
@@ -1072,6 +1146,125 @@ Protocol.prototype.bindIO = function bindIO () {
...
+				userIdArr: userIdArr,
+				regionId: regionId
+			}, function (err, data) {
+				protocol.updateProtectedRegions(socket.room);
+				callback(err, data);
+			});
+		});
+
+		socket.on("removeUsersFromMyProtectedRegion", function (userIdArr, regionId, callback) {
+			if(!socket.userid)			{
```
Comment (Squarific) [inline server/realTime/scripts/Network.js]: More random spaces :p
